tools/Google.GCloud.Tools.GenerateSnippetMarkdown/Program.cs

```diff
@@ -0,0 +1,130 @@
+    public sealed class Program
+    {
+        private static readonly Regex StartSnippetTag = new Regex(@"^\s*// \<(?<name>[\w\.]+)>\s*$", RegexOptions.Compiled);
+
+        private static void Main(string[] args)
+        {
+            string root = DetermineRootDirectory();
+            if (root == null)
+            {
+                Console.WriteLine("Unable to determine root directory. Please run within gcloud-dotnet repository.");
+                return;
+            }
+
+            string snippets = Path.Combine(root, "snippets");
+            if (!Directory.Exists(snippets))
+            {
+                Console.WriteLine($"Snippets directory {snippets} doesn't exist. Aborting.");
+                return;
+            }
+
+            string metadata = Path.Combine(root, "docs", "obj", "api");
+            if (!Directory.Exists(metadata))
+            {
+                Console.WriteLine($"Metadata directory {metadata} doesn't exist. Aborting.");
+                return;
+            }
+
+            string output = Path.Combine(root, "docs", "extra");
+            if (!Directory.Exists(output))
+            {
+                Console.WriteLine($"Output directory {output} doesn't exist. Aborting.");
+                return;
+            }
+            GenerateSnippetUids(snippets, metadata, Path.Combine(output, "snippets.md"));
+        }
+
+        /// <summary>
+        /// Generates a markdown file 
+        /// </summary>
+        /// <param name="snippetsDir"></param>
+        /// <param name="metadataDir"></param>
+        /// <param name="outputFile"></param>
+        private static void GenerateSnippetUids(string snippetsDir, string metadataDir, string outputFile)
+        {
+            // TODO: Use a YAML parser
+            var uids = new HashSet<string>(
+                from yaml in Directory.GetFiles(metadataDir, "Google*.yml")
+                from line in File.ReadLines(yaml)
+                let trimmed = line.Trim()
+                where line.StartsWith("- uid: ")
+                select line.Substring("- uid: ".Length)).ToList();
+
+            using (var outputMarkdown = File.CreateText(outputFile))
+            {
+                // Search every line, in every source file, in every project.
+                // If we find a snippet, e.g. // <snippetName>, then find all the UIDs from YAML
+                // files that are for that method name, scoped by the project/source names.
+                var query = from project in Directory.GetDirectories(snippetsDir)
+                            // Path.GetFileName just takes the last part of the name; it doesn't know that it's a directory.
+                            let projectName = TrimSuffix(Path.GetFileName(project), ".Snippets")
+                            from sourceFile in Directory.GetFiles(project, "*.cs")
+                            let sourceRelativeToOutput = $"../../snippets/{Path.GetFileName(project)}/{Path.GetFileName(sourceFile)}"
```

Comment:
Well, not with Combine. We basically need a way of getting a relative path - definitely using forward slashes - from outputFile to snippetsDir. Will have a think.